### ISSUE TITLE
Assignment in initializer between two uninitialized variables results in an initialized variable

### ISSUE THREAD

**Issue by GoogleCodeExporter** [typetools/checker-framework#345]
```
What steps will reproduce the problem?
Run the NullnessChecker on the following code snippet (also attached):
public class Test {
    String f1;
    String f2;

    {
        f1 = f2;
        f2 = f1;
        f2.toString();   //Null pointer exception here
    }
    
    public static void main(String [] args) {
        Test a = new Test();
    }
}

What is the expected output? What do you see instead?
The first assignment should issue a warning since it would expect an 
initialized value.  EIther that, or the field f1 should not be considered 
initialized after the assignment.


What version of the product are you using? On what operating system?
Checker Framework 1.8.4
```

Original issue reported on code.google.com by `Jonathan...@gmail.com` on 8 Aug 2014 at 7:32

Attachments:
* [Test.java](https://storage.googleapis.com/google-code-attachments/checker-framework/issue-345/comment-0/Test.java)

**Comment by smillst** [comment]
Test cases are checker/tests/nullness/AssignmentDuringInitialization.java and checker/tests/nullness/Issue345.java

**Comment by smillst** [comment]
The errors are correctly issued when caching in the AnnotatedTypeFactory is turned off:
````
javacheck -processor nullness AssignmentDuringInitialization.java -AatfDoNotCache
AssignmentDuringInitialization.java:16: error: [assignment.type.incompatible] incompatible types in assignment.
        f1 = f2;
        ^
  found   : @Initialized @Nullable String
  required: @Initialized @NonNull String
AssignmentDuringInitialization.java:16: error: [assignment.type.incompatible] incompatible types in assignment.
        f1 = f2;
             ^
  found   : @Initialized @Nullable String
  required: @Initialized @NonNull String
AssignmentDuringInitialization.java:23: error: [assignment.type.incompatible] incompatible types in assignment.
        f3 = f4;
        ^
  found   : @Initialized @Nullable String
  required: @Initialized @NonNull String
AssignmentDuringInitialization.java:23: error: [assignment.type.incompatible] incompatible types in assignment.
        f3 = f4;
             ^
  found   : @Initialized @Nullable String
  required: @Initialized @NonNull String
````
(The -AatfDoNotCache is currently only available in the caching branch.)

**Comment by smillst** [comment]
````
javacheck -processor nullness AssignmentDuringInitialization.java -AatfDoNotCache
AssignmentDuringInitialization.java:16: error: [assignment.type.incompatible] incompatible types in assignment.
        f1 = f2;
        ^
  found   : @Initialized @Nullable String
  required: @Initialized @NonNull String
````
The found type is incorrect since `f2` is not initialized.  The type of `f2` should be `@UnknownInitialization`.

(PR #605 does not address this issue.)

**Comment by smillst** [comment]
Fixed by PR #605
